agent/lib_aws_sdk_php.c

```diff
@@ -295,6 +302,194 @@ void nr_lib_aws_sdk_php_sqs_parse_queueurl(
+      if (nr_php_is_zval_non_empty_string(uri)) {
+        external_params.uri = Z_STRVAL_P(uri);
+      }
+    }
+    
+  }
+  nr_segment_external_end(&external_segment, &external_params);
+  nr_free(cloud_attrs.cloud_resource_id);
+}
+
+/* This stores the compiled regex to parse AWS ARNs. The compilation happens when
+ * it is first needed and is destroyed in mshutdown
+ */
+static nr_regex_t* aws_arn_regex;
+
+static void nr_aws_sdk_compile_regex(void) {
+  aws_arn_regex = nr_regex_create(AWS_LAMBDA_ARN_REGEX, 0, 0);
+}
+
+void nr_aws_sdk_mshutdown(void) {
+  nr_regex_destroy(&aws_arn_regex);
+}
+
+void nr_aws_sdk_lambda_client_invoke_parse_args(NR_EXECUTE_PROTO, nr_segment_cloud_attrs_t* cloud_attrs) {
+  zval* call_args = nr_php_get_user_func_arg(2, NR_EXECUTE_ORIG_ARGS);
+  zval* this_obj = getThis();//NR_PHP_USER_FN_THIS();
+  char* arn = NULL;
+  char* function_name = NULL;
+  char* region = NULL;
+  zval* region_zval = NULL;
+  char* qualifier = NULL;
+  char* accountID = NULL;
+  bool using_account_id_ini = false;
+
+  /* verify arguments */
+  if (!nr_php_is_zval_valid_array(call_args)) {
+    return;
+  }
+  zval* lambda_args = nr_php_zend_hash_index_find(Z_ARRVAL_P(call_args), 0);
+  if (!nr_php_is_zval_valid_array(lambda_args)) {
+    return;
+  }
+  zval* lambda_name = nr_php_zend_hash_find(Z_ARRVAL_P(lambda_args), "FunctionName");
+  if (!nr_php_is_zval_non_empty_string(lambda_name)) {
+    return;
+  }
+
+  /* Ensure regex exists */
+  if (NULL == aws_arn_regex) {
+    nr_aws_sdk_compile_regex();
+  }
+
+  /* Extract all information possible from the passed lambda name via regex */
+  nr_regex_substrings_t* matches =
+      nr_regex_match_capture(aws_arn_regex,
+                             Z_STRVAL_P(lambda_name),
+                             Z_STRLEN_P(lambda_name));
+  function_name = nr_regex_substrings_get_named(matches, "functionName");
+  accountID = nr_regex_substrings_get_named(matches, "accountId");
+  region = nr_regex_substrings_get_named(matches, "region");
+  qualifier = nr_regex_substrings_get_named(matches, "qualifier");
+
+  /* supplement missing information with API calls */
+  if (nr_strempty(function_name)) {
+    /*
+     * Cannot get the needed data. Function name is required in the
+     * argument, so this won't happen in normal operation
+     */
+    nr_free(function_name);
+    nr_free(accountID);
+    nr_free(region);
+    nr_free(qualifier);
+    nr_regex_substrings_destroy(&matches);
+    return;
+  }
+  if (nr_strempty(accountID)) {
+    nr_free(accountID);
+    accountID = NRINI(aws_account_id);
+    using_account_id_ini = true;
+  }
+  if (nr_strempty(region)) {
+    zend_class_entry* base_class = NULL;
+    if (NULL != execute_data->func && NULL!= execute_data->func->common.scope) {
+       base_class = execute_data->func->common.scope;
+    }
+    region_zval
+      = nr_php_get_zval_object_property_with_class(this_obj, base_class, "region");
```

Comment:
Is this construct correct? Take a look at [this example](https://github.com/newrelic/newrelic-php-agent/blob/R11.6/agent/lib_guzzle6.c#L411-L420). The first argument to `nr_php_get_zval_object_property_with_class` is the object to extract the property from, and the second argument is the class entry for the context from which the property should be extracted. What is this code trying to do?

Reply:
`LambdaClient` extends `AwsClient`. `region` exists on the `AwsClient`, not the `LambdaClient`. So if we wish to extract `region` from `this` (which is a `LambdaClient`), we must indicate the base class.

Reply:
If the intent is to look for the `"region"` property either in `this` (which is `Aws\Lambda\LambdaClient`) and then fall back to looking for it in `this`'s base class (which is `Aws\AwsClient`), then wouldn't it be safer to get the `base_class` like this: `base_class = nr_php_find_class("aws\\awsclient");`?

Reply:
The best strategy to ensure that the agent will be able to retrieve the region is to call `getRegion` method on `this` like this: `region = nr_php_call(this, "getRegion");` It doesn't rely on any assumptions but rather on publicly available API of AWS SDK for PHP.